### PR TITLE
Adding user_id to metrics

### PR DESCRIPTION
**Pull request by josemazo**
Issue: https://app.clubhouse.io/cartoteam/story/58519/track-user-in-metrics

### CODE REVIEW

File: cartoframes/utils/metrics.py [hunk 1]
```diff
@@ -79,11 +80,14 @@ def build_metrics_data(event_name):
         'runtime_env': get_runtime_env()
     }
 
+    metrics_data.update(extra_metrics_data)  # Compatible with old Python versions
+    return metrics_data
+
 
 @silent_fail
-def post_metrics(event_name):
+def post_metrics(event_name, extra_metrics_data={}):
```
[the remark this answers is not in the document]
Comment (josemazo) [reply cartoframes/utils/metrics.py]: Fixed! Cool tool :kissing_heart:

File: cartoframes/utils/metrics.py [hunk 2]
```diff
@@ -68,8 +69,8 @@ def check_valid_metrics_uuid(metrics_config):
     return metrics_config is not None and is_uuid(metrics_config.get(UUID_KEY))
 
 
-def build_metrics_data(event_name):
-    return {
+def build_metrics_data(event_name, extra_metrics_data={}):
```
[the remark this answers is not in the document]
Comment (josemazo) [reply cartoframes/utils/metrics.py]: Fixed!

File: cartoframes/utils/metrics.py [hunk 3]
```diff
@@ -79,11 +80,17 @@ def build_metrics_data(event_name):
         'runtime_env': get_runtime_env()
     }
 
+    if not isinstance(extra_metrics_data, dict):
```
Comment (Jesus89) [inline cartoframes/utils/metrics.py]: We need @cmongut @ilbambino for the definition of the metrics structure. Maybe they require "user_id" instead of a field called "extra", so I think we could merge both dictionaries (metrics_data + extra_metrics_data)

File: cartoframes/utils/metrics.py [hunk 4]
```diff
@@ -93,7 +100,13 @@ def decorator_func(func):
         @functools.wraps(func)
         def wrapper_func(*args, **kwargs):
             result = func(*args, **kwargs)
-            post_metrics(event_name)
+
+            credentials = get_parameter_from_decorator('credentials', func, *args, **kwargs)
```
Comment (Jesus89) [inline cartoframes/utils/metrics.py]: I would extract this to a function `build_extra_metrics_data`

File: cartoframes/utils/utils.py [hunk 5]
```diff
@@ -334,7 +335,8 @@ def is_uuid(text):
 def get_credentials(credentials=None):
     from ..auth import defaults
     _credentials = credentials or defaults.get_default_credentials()
-    check_credentials(_credentials)
+    if credentials is not None:
```
Comment (Jesus89) [inline cartoframes/utils/utils.py]: Why is this added? If `credentials` is None but you have `default_credentials` it should call `check_credentials` too.

File: cartoframes/utils/utils.py [hunk 6]
```diff
@@ -524,3 +526,18 @@ def fn(*args, **kw):
         except Exception:
             pass
     return fn
+
+
+def get_parameter_from_decorator(parameter_name, func, *args, **kwargs):
```
Comment (Jesus89) [inline cartoframes/utils/utils.py]: Cool!

File: cartoframes/utils/utils.py (hunk 5 above)
Comment (josemazo) [reply cartoframes/utils/utils.py]: There are some tests that have no credentials configured, and `check_credentials` raises an exception.

File: cartoframes/utils/metrics.py (hunk 3 above)
Comment (josemazo) [reply cartoframes/utils/metrics.py]: Yes, it isn't defined. But the extra part is almost an error. The just uploaded commit it only adds the `extra` key if the `extra_metrics_data` is not a `dict`.

**Comment by josemazo** [comment]
Ready for CR again!

File: cartoframes/utils/utils.py (hunk 5 above)
Comment (Jesus89) [reply cartoframes/utils/utils.py]: Aha, in that case, I would try to pass a real Credential instance to those specific tests.

Comment (josemazo) [reply cartoframes/utils/utils.py]: The failing tests were the `Map` ones, and the `Map` class doesn't accept a `Credentials` parameter, so I've used the `set_default_credentials` function and created a new `unset_default_credentails` becuase if not, the `test_source_no_credentials` test fails because it has the default credentials in the global variables.

**Comment by josemazo** [comment]
Ready for re-CR!

File: cartoframes/utils/metrics.py (hunk 3 above)
Comment (Jesus89) [reply cartoframes/utils/metrics.py]: I wouldn't add the `extra` key until it's required from `Product`, so I think we can remove the extra_metrics_data as  != dict and return just {}

**Comment by josemazo** [comment]
Adding PEP448! :muscle:
